[tayxuenye] iP

tayxuenye opened the pull request:
### Optimus

- text-based
- easy to learn
- ~~fast~~ **super fast** to use

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

1. download it from [[here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)](url)
2. double-click it
3. add your tasks 😉

Features:
- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Managing events

If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

tayxuenye committed:
changed name of bot from Duke to Optimus

tayxuenye committed:
Bot now greets user and exits

tayxuenye committed:
echos commands entered by user, and exits when user types bye

tayxuenye committed:
removed lines for easier coding

tayxuenye committed:
added Duke Level-2 functionalities add and list

tayxuenye committed:
Added mark as done feature for Level-3

tayxuenye committed:
comment formatting changes

tayxuenye committed:
changed addTask method to take in String instead of Task. implemented support for adding 3 types of tasks.

tayxuenye committed:
Edited listTasks and markTaskAsDone methods to make sure the type of task is also printed

tayxuenye committed:
Added some error handling. Accidentally started on Level 5 first

tayxuenye committed:
added test cases for semi-auto testing. Accidentally started on Level-5 first

tayxuenye committed:
Added error handling for todo and blah

tayxuenye committed:
changed taskList from type Array to type ArrayList. Implemented delete task function in Optimus class

tayxuenye committed:
Implement loadTasks and edit style with ChatGPT input

- Implemented the loadTasks method to read tasks from a file.\n- Edited code style for better readability and adherence to CS2103T guidelines.\n- Acknowledged usage of ChatGPT for code suggestions and inspiration.

tayxuenye committed:
Add saveTasks method

Implemented saveTasks method to save tasks to file. \nAdded calls to saveTasks after marking, deleting, and adding tasks. \n Followed ChatGPT's suggestions and applied CS2103T style guide

tayxuenye committed:
Merge branch-Level-7 into master after implementing Level 7

tayxuenye committed:
Level-8: Add support for dates in Deadline and Event tasks

Implemented date handling for Deadline and Event tasks using java.time.LocalDate.\n The chatbot now accepts deadlines and events with dates in the format yyyy-mm-dd and prints them in the format MMM d yyyy.\n Added error handling for invalid date formats

tayxuenye committed:
Merge branch 'branch-Level-8'

tayxuenye committed:
Refactor code to abstract out Ui, Storage, Parser, and TaskList

Ensured all functionalities from Level 8 are retained. \nImproved error handling in TaskList for marking and deleting tasks to prevent IndexOutOfBoundsException. \n Refactoted code to abstract out TaskList, Ui, Storage, and Parser classes

tayxuenye committed:
Refactor code into optimus package

tayxuenye committed:
Merge A-Packages into master

tayxuenye committed:
Merge branch 'add-gradle-support' of https://github.com/tayxuenye/ip into add-gradle-support

tayxuenye committed:
Set up Gradle and successfully configured the build

tayxuenye committed:
Merge branch 'add-gradle-support'

tayxuenye committed:
Create executable JAR file and add Gradle shadowJar support

tayxuenye committed:
Implement Level-9: Add Find Feature to Search Tasks by Keyword

- Added the ind command to allow users to search tasks by keyword.

- Updated Ui class with a indTasks method to display matching tasks.

- Ensured search functionality is integrated with the task list.

- Added error handling for invalid find commands.

tayxuenye committed:
Merge branch 'branch-Level-9'

tayxuenye committed:
Add JUnit tests for TaskList for addTask and markTaskAsDone methods

tayxuenye committed:
Add additional comments to give credit to ChatGPT

tayxuenye committed:
Changes to comments

tayxuenye committed:
Changes to comments

tayxuenye committed:
Integrate GUI from JavaFX tutorial into Optimus

Previously, it was just a text-based interface. Now it is much more usable and aesthetically pleasing, and users will have a text input to type.

tayxuenye committed:
make style edits

tayxuenye committed:
Edit style to match style check requirements

yuanchengg commented on src/main/java/optimus/Main.java:
```diff
@@ -0,0 +1,54 @@
+ * Adapted from CS2103T JavaFX tutorial -reused
+ */
+public class Main extends Application {
+
```
maybe can remove this line here. not sure if there is supposed to be a line here

yuanchengg commented on src/main/java/optimus/OptimusException.java:
```diff
@@ -0,0 +1,16 @@
+     *
+     * @param msg The detail message for the exception.
+     */
+    public OptimusException(String msg) {
```
like how you used egyptian style coding in your code 😄

yuanchengg commented on src/main/java/optimus/Ui.java:
```diff
@@ -0,0 +1,96 @@
+        this.mainWindow = mainWindow;
+    }
+
+    // Method to show messages to the user in the chatbox
```
should the first sentence start in the form Returns ..., Sends ..., Adds ... etc. (not Return or Returning etc.)

yuanchengg commented on src/main/java/optimus/Main.java:
```diff
@@ -0,0 +1,54 @@
+package optimus;
+import java.io.IOException;
```
should there be a line here?

yuanchengg commented on src/test/java/optimus/TaskListTest.java:
```diff
@@ -0,0 +1,123 @@
+package optimus;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
```
like how you order the imports nicely !! 😄

yuanchengg reviewed (commented):
LGTM. coding standards are followed very closely and well. love it!

rubinnn commented on src/main/java/optimus/DialogBox.java:
```diff
@@ -0,0 +1,67 @@
+package optimus;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+import javafx.scene.shape.Circle;
```
Good job with grouping the imports

rubinnn commented on src/main/java/optimus/Optimus.java:
```diff
@@ -0,0 +1,77 @@
+    public void run(String userInput) {
+        try {
+            String[] command = Parser.parseCommand(userInput);
+            if (command[0].equals("bye")) { // Exit the program
+                ui.sayBye();
+                Platform.exit();
+            } else if (command[0].equals("list")) { // List all tasks
+                ui.listTasks(taskList);
+            } else if (command[0].equals("mark")) { // Mark a task as done
+                taskList.markTaskAsDone(Integer.parseInt(command[1]) - 1);
+                storage.saveTasks(taskList);
+            } else if (command[0].equals("delete")) { // Delete a task
+                if (command.length != 2) { // Check if command is in format like "delete 2"
+                    throw new OptimusException("Invalid format. Use 'delete [task number]'.");
+                }
+                taskList.delete(Integer.parseInt(command[1]) - 1);
+                storage.saveTasks(taskList);
+            } else if (command[0].equals("todo") || command[0].equals("deadline") || command[0].equals("event")) {
+                // Add a task
+                taskList.addTask(userInput);
+                storage.saveTasks(taskList);
+            } else if (command[0].equals("find")) { // Find tasks which contain keyword and print them out
+                if (command.length != 2) { // Check if command is in format like "find book"
+                    throw new OptimusException("Invalid format. Use 'find [keyword]'.");
+                }
+                ui.findTasks(userInput, taskList);
+            } else { // Handle invalid command
+                ui.showExampleCommand();
+                throw new OptimusException("Invalid command.");
+            }
```
Should a switch statement be used?

rubinnn commented on src/main/java/optimus/Task.java:
```diff
@@ -0,0 +1,54 @@
+package optimus;
```
Good job with the clean code.

rubinnn commented on src/main/java/optimus/Ui.java:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * Greets the user and provides instructions for using Optimus.
+     */
+    public void greetUser() {
+        String greeting = "Hello! I'm Optimus.\n"
+                + "What can I do for you?\n"
+                + "You need to start your input with either todo, deadline, or event.\n"
+                + "For example:\n"
+                + "todo borrow book\n"
+                + "deadline return book /by 2019-12-02\n"
+                + "event project meeting /from 2019-12-02 /to 2019-12-03\n"
+                + "You can also do e.g.\n"
+                + "mark 2\n"
+                + "delete 1\n"
+                + "list";
+        showToUser(greeting);
+    }
```
Good job extracting out the responses for easier access to the commands.

rubinnn commented on src/main/java/optimus/Storage.java:
```diff
@@ -0,0 +1,124 @@
+     * @return The list of tasks loaded from the file.
+     * @throws IOException If an I/O error occurs while loading the file.
+     */
+    public ArrayList<Task> loadTasks() throws IOException {
```
Good naming conventions overall in the code. Good job

rubinnn reviewed (commented):
LGTM. Just some nits to fix.

tayxuenye committed:
Create new branch level 10 to satisfy iP requirements

tayxuenye committed:
Parser and TaskList classes: Add assertion for input validation

The parseCommand method in the Parser class currently does not check if the user input is empty before splitting the string. Additionally, the markTaskAsDone and delete methods in TaskList do not validate if the provided index is within bounds before proceeding.

tayxuenye committed:
Refactor code to adhere to CS2103T coding standards

I have examined and refactored the code to follow the CS2103T coding standards. This involved adding comments to assertions to make sure assumptions are well justified.

tayxuenye committed:
Modify findTasks to allow BetterSearch by supporting different case characters

tayxuenye committed:
Merge pull request #2 from tayxuenye/branch-A-CodeQuality

Refactor code to adhere to CS2103T coding standards

tayxuenye committed:
Merge branch 'master' of https://github.com/tayxuenye/ip

tayxuenye committed:
Refactor project to conform to iP requirements

tayxuenye committed:
Refactor project to conform to iP requirements

tayxuenye committed:
Refactor project to conform to iP requirements

tayxuenye committed:
Refactor project to conform to iP requirements

tayxuenye committed:
Refactor project to conform to iP requirements

tayxuenye committed:
Add UI screenshot and update user guide

tayxuenye committed:
Fix bug in README

tayxuenye committed:
Fix bug in README

tayxuenye committed:
Fix bug in README

tayxuenye committed:
Fix bug in README

tayxuenye committed:
Update code to satisfy style check

tayxuenye committed:
Fix bugs when using jar file

tayxuenye committed:
Edit code reuse citation

tayxuenye committed:
Update README to include Optimus.png
